tests/core/test_optimizers_schedulers.py

```diff
@@ -616,6 +617,48 @@ def test_CosineAnnealing(self):
 
         assert final_lr == self.MIN_LR
 
+    # Noam scheduler should decay past MAX_STEPS - run two schedulers in parallel to test it
+    @pytest.mark.unit
+    def test_NoamAnnealing(self):
+        model = TempModel()
+        opt_cls = optim.get_optimizer('novograd')
+        opt1 = opt_cls(model.parameters(), lr=self.INITIAL_LR)
+        opt2 = opt_cls(model.parameters(), lr=self.INITIAL_LR)
+
+        # No warmup case
+        policy1 = optim.lr_scheduler.NoamAnnealing(opt1, d_model=self.D_MODEL, max_steps=self.MAX_STEPS, min_lr=self.MIN_LR)
+        policy2 = optim.lr_scheduler.NoamAnnealing(opt2, d_model=self.D_MODEL, max_steps=self.MAX_STEPS*2, min_lr=self.MIN_LR)
+        initial_lr = policy1.get_last_lr()[0]
+
+        assert initial_lr == self.D_MODEL**(-0.5) * self.INITIAL_LR
+
+        for i in range(self.MAX_STEPS * 2):
+            assert self.MIN_LR < policy1.get_last_lr()[0] <= self.INITIAL_LR
+            assert policy1.get_last_lr()[0] == policy2.get_last_lr()[0]
+            opt1.step()
+            opt2.step()
+            policy1.step()
+            policy2.step()
+
+        # Warmup steps available
+        policy1 = optim.lr_scheduler.NoamAnnealing(opt1, d_model=self.D_MODEL, warmup_steps=5, max_steps=self.MAX_STEPS, min_lr=self.MIN_LR)
+        policy2 = optim.lr_scheduler.NoamAnnealing(opt2, d_model=self.D_MODEL, warmup_steps=5, max_steps=self.MAX_STEPS*2, min_lr=self.MIN_LR)
+        initial_lr = policy1.get_last_lr()[0]
+
+        assert initial_lr < self.INITIAL_LR
+
+        for i in range(self.MAX_STEPS * 2):
+            if i <= 5:
+                assert policy1.get_last_lr()[0] <= self.INITIAL_LR
+            else:
+                assert self.MIN_LR < policy1.get_last_lr()[0] < self.INITIAL_LR
+                assert policy1.get_last_lr()[0] == policy2.get_last_lr()[0]
+
+            opt1.step()
+            opt2.step()
+            policy1.step()
+            policy2.step()
+
```

Comment:
nit: it would be easier to maintain unit tests if decoupling warmup and no_warmup as two separate functions, for example, 

```
def test_NoamAnnealing_without_warmup(self):
    pass

def test_NoamAnnealing_with_warmup(self):
    pass
```